agent/lib/kontena/workers/node_info_worker.rb

```diff
@@ -245,6 +264,93 @@ def calculate_container_time(container)
     end
 
+    # @param [Array<Vmstat::Cpu>] prev_cpus
+    # @param [Array<Vmstat::Cpu>] current_cpu
+    # @return [Hash] { :num_cores, :system, :user, :idle }
+    def calculate_cpu_usage(prev_cpus, current_cpus)
+      result = {
+        num_cores: prev_cpus.size,
+        system: 0.0,
+        user: 0.0,
+        nice: 0.0,
+        idle: 0.0
+      }
+
+      prev_cpus.zip(current_cpus).map { |prev_cpu, current_cpu|
+        system_ticks = current_cpu.system - prev_cpu.system
+        user_ticks = current_cpu.user - prev_cpu.user
+        nice_ticks = current_cpu.nice = prev_cpu.nice
+        idle_ticks = current_cpu.idle - prev_cpu.idle
+
+        total_ticks = (system_ticks + user_ticks + nice_ticks + idle_ticks).to_f
+
+        {
+          system: (system_ticks / total_ticks) * 100.0,
+          user: (user_ticks / total_ticks) * 100.0,
+          nice: (nice_ticks / total_ticks) * 100.0,
+          idle: (idle_ticks / total_ticks) * 100.0
+        }
+      }.inject(result) { |memo, cpu_core|
+        memo[:system] += cpu_core[:system]
+        memo[:user] += cpu_core[:user]
+        memo[:nice] += cpu_core[:nice]
+        memo[:idle] += cpu_core[:idle]
+        memo
+      }
+    end
+
+    # @param [Array<Vmstat::NetworkInterface>] prev_interfaces
+    # @param [Array<Vmstat::NetworkInterface>] current_interfaces
+    # @param [Number] interval_seconds
+    # @return [Hash]
+    def calculate_network_traffic(prev_interfaces, current_interfaces, interval_seconds)
+      internal_interfaces = current_interfaces.select { |iface|
+        iface.name.to_s == "weave" or iface.name.to_s.start_with?("vethwe")
+      }
+
+      external_interfaces = current_interfaces.select { |iface|
+        iface.name.to_s == "docker0"
+      }
+
+      {
+        internal: calculate_interface_traffic(prev_interfaces, internal_interfaces, interval_seconds),
+        external: calculate_interface_traffic(prev_interfaces, external_interfaces, interval_seconds)
+      }
+    end
+
+    def calculate_interface_traffic(prev_interfaces, current_interfaces, interval_seconds)
+      results = {
+        interfaces: [],
+        rx_bytes: 0,
+        prev_rx_bytes: 0,
+        tx_bytes: 0,
+        prev_tx_bytes: 0
+      }
+
+      results = current_interfaces.inject(results) { |result, iface|
+        result[:interfaces] << iface.name
+        result[:rx_bytes] += iface.in_bytes
+        result[:tx_bytes] += iface.out_bytes
+
+        prev_iface = prev_interfaces.select { |x| x.name == iface.name }
+
+        if prev_iface.size > 0
+          result[:prev_rx_bytes] += prev_iface[0].in_bytes
+          result[:prev_tx_bytes] += prev_iface[0].out_bytes
+        end
+
+        result
+      }
+
+      results[:rx_bytes_per_second] = ((results[:rx_bytes] - results[:prev_rx_bytes]).to_f / interval_seconds).round
+      results[:tx_bytes_per_second] = ((results[:tx_bytes] - results[:prev_tx_bytes]).to_f / interval_seconds).round
+
+      results.delete(:prev_rx_bytes)
```

Comment:
Instead of having a hash that changes form within the same function, just use separate variables and then return a new hash?

```ruby
{
  interfaces: interfaces,
  rx_bytes: total_rx_bytes,
  tx_bytes: total_tx_bytes,
  rx_bytes_per_second: ....
  tx_bytes_per_second: ....
}
```

IMO easier to figure out what's actually getting returned.